### DISCUSSION TITLE
Pulumi Code Destructurization

### DISCUSSION THREAD

**Discussion by kedar9696** [pulumi/pulumi#16685] [Automation API]
Can we desctructure a pulumi cod like terraform
eg . main.tf , varibale.tf, provider.tf, output.tf like this

**Comment by smithrobs** [comment]
Pulumi is quite flexible - you can separate your code as you deem appropriate. See [Organizing your project code](https://www.pulumi.com/docs/using-pulumi/organizing-projects-stacks/#organizing-your-project-code) and more generally the [IaC Recommended Practices](https://www.pulumi.com/blog/iac-recommended-practices-code-organization-and-stacks/) blog series.

In your specific case, you could do something like (assuming Python):
* `main.tf` -> `__main__.py` (`import`s the following files and creates the resources in your Stack)
* `variables.tf` -> `config.py` (could contain all your Pulumi config gathered from `pulumi.Config()`)
* `provider.tf` -> `provider.py` (could contain [explicit providers](https://www.pulumi.com/docs/concepts/resources/providers/#explicit-provider-configuration))
* `output.tf` -> `output.py` (output whatever you need via `pulumi.export`)
* etc.

Do what works for you! I would not recommend this approach as a "Pulumi Code Best Practice". "Small" projects tend to be single-file. Larger projects tend to get destructured into [Component Resources](https://www.pulumi.com/docs/concepts/resources/components/) as functional areas or separate/dependent [Stack References](https://www.pulumi.com/learn/building-with-pulumi/stack-references/).

**Reply by kedar9696** [reply]
Can we restructurize code written in YAML language. ?

**Reply by smithrobs** [reply]
if you mean you want to have multiple Pulumi YAML files - it's not directly supported, but @AaronFriel has an [example/workaround](https://github.com/pulumi/pulumi-yaml/issues/227#issuecomment-1930589084)
